src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -705,24 +743,25 @@ private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ArraySeg
 
                     if (_isServer)
                     {
-                        _receivedMaskOffsetOffset = ApplyMask(_receiveBuffer, _receiveBufferOffset, header.Mask, _receivedMaskOffsetOffset, bytesToCopy);
+                        _receivedMaskOffsetOffset = ApplyMask(_receiveBuffer.Span.Slice(_receiveBufferOffset, bytesToCopy), header.Mask, _receivedMaskOffsetOffset);
                     }
-                    Buffer.BlockCopy(_receiveBuffer, _receiveBufferOffset, payloadBuffer.Array, payloadBuffer.Offset, bytesToCopy);
+                    _receiveBuffer.Span.Slice(_receiveBufferOffset, bytesToCopy).CopyTo(payloadBuffer.Span.Slice(0, bytesToCopy));
                     ConsumeFromBuffer(bytesToCopy);
                     header.PayloadLength -= bytesToCopy;
 
                     // If this a text message, validate that it contains valid UTF8.
                     if (header.Opcode == MessageOpcode.Text &&
-                        !TryValidateUtf8(new ArraySegment<byte>(payloadBuffer.Array, payloadBuffer.Offset, bytesToCopy), header.Fin, _utf8TextState))
+                        !TryValidateUtf8(payloadBuffer.Span.Slice(0, bytesToCopy), header.Fin, _utf8TextState))
```

Comment:
Future candidate for `Utf8Utility.IsWellFormedUtf8String`? https://github.com/dotnet/corefxlab/pull/1881 or alternative routine for benchmarking against. 

/cc @GrabYourPitchforks

Reply:
Thanks for pointing out this use case! Solidifies my belief that I need to add a streaming API to the validator.

Reply:
https://github.com/dotnet/corefxlab/pull/1918 should provide a centralized drop-in replacement for this logic.